tests/waku_relay/test_protocol.nim

```diff
@@ -0,0 +1,1302 @@
+  var node {.threadvar.}: WakuRelay
+  var remotePeerInfo {.threadvar.}: RemotePeerInfo
+  var peerId {.threadvar.}: PeerId
+  var contentTopic {.threadvar.}: ContentTopic
+  var pubsubTopic {.threadvar.}: PubsubTopic
+  var pubsubTopicSeq {.threadvar.}: seq[PubsubTopic]
+  var testMessage {.threadvar.}: string
+  var wakuMessage {.threadvar.}: WakuMessage
+
+  asyncSetup:
+    messageSeq = @[]
+    handlerFuture = newPushHandlerFuture()
+    simpleFutureHandler = proc(
+      topic: PubsubTopic, msg: WakuMessage
+    ): Future[void] {.async, closure, gcsafe.} =
+      messageSeq.add((topic, msg))
+      handlerFuture.complete((topic, msg))
+
+    switch = newTestSwitch()
+    peerManager = PeerManager.new(switch)
+    node = await newTestWakuRelay(switch)
+
+    testMessage = "test-message"
+    contentTopic = DefaultContentTopic
+    pubsubTopic = DefaultPubsubTopic
+    pubsubTopicSeq = @[pubsubTopic]
+    wakuMessage = fakeWakuMessage(testMessage, pubsubTopic)
+
+    await allFutures(switch.start(), node.start())
+
+    remotePeerInfo = switch.peerInfo.toRemotePeerInfo()
+    peerId = remotePeerInfo.peerId
+
+  asyncTeardown:
+    await allFutures(switch.stop(), node.stop())
+
+  suite "Subscribe":
+    asyncTest "Publish without Subscription":
+      # When publishing a message without being subscribed
+      discard await node.publish(pubsubTopic, wakuMessage)
+
+      # Then the message is not published
+      check:
+        not await handlerFuture.withTimeout(3.seconds)
+
+    asyncTest "Publish with Subscription (Network Size: 1)":
+      # When subscribing to a Pubsub Topic
+      discard node.subscribe(pubsubTopic, simpleFutureHandler)
+
+      # Then the node is subscribed
+      check:
+        node.isSubscribed(pubsubTopic)
+        node.subscribedTopics == pubsubTopicSeq
+
+      # When publishing a message
+      discard await node.publish(pubsubTopic, wakuMessage)
+
+      # Then the message is published
+      assert (await handlerFuture.withTimeout(3.seconds))
+      let (topic, msg) = handlerFuture.read()
+      check:
+        topic == pubsubTopic
+        msg == wakuMessage
+    
+    asyncTest "Pubsub Topic Subscription (Network Size: 2, only one subscribed)":
+      # Given a second node connected to the first one
+      let 
+        otherSwitch = newTestSwitch()
+        otherNode = await newTestWakuRelay(otherSwitch)
+
+      await allFutures(otherSwitch.start(), otherNode.start())
+      let otherRemotePeerInfo = otherSwitch.peerInfo.toRemotePeerInfo()
+      check await peerManager.connectRelay(otherRemotePeerInfo)
+
+      var otherHandlerFuture = newPushHandlerFuture()
+      proc otherSimpleFutureHandler(topic: PubsubTopic, message: WakuMessage) {.async, gcsafe.} =
+        otherHandlerFuture.complete((topic, message))
+
+      # When subscribing the second node to the Pubsub Topic
+      discard otherNode.subscribe(pubsubTopic, otherSimpleFutureHandler)
+
+      # Then both nodes are subscribed
```

Comment:
Might be an error? isn't here only on node subscribed?

Reply:
Indeed, too much copy-pasting, thanks.